cyblog.md

```diff
@@ -0,0 +1,75 @@
+
+Unfortunately, while `.apply` gives a lot of power and flexibility, it is also somewhat verbose and clunky for describing such a simple concept.  Fortunately, we found a way to improve the situation greatly.  This realization came about from great work done in our Python package.  First, before diving into the low level details, let's first see what the above code would now look like in 0.17.0:
+
+```ts
+// certCertificate and zone created as above
+
+const certValidation = new aws.route53.Record("cert_validation", {
+    records: [certCertificate.domainValidationOptions[0].resourceRecordValue],
+    ttl: 60,
+    type: certCertificate.domainValidationOptions[0].resourceRecordType,
+    zoneId: zone.id,
+});
+```
+
+That's a lot nicer than before!  The following improvements happened:
+1. there's no more lambda!
+2. there's no need for a repetive lambda parameter (which might conflict with some other name in scope).
+3. the code *is* just pure simple *idiomatic* JavaScript/TypeScript that clearly conveys its intent.
+
+Importantly, no information has been lost here.  The exact same dependency information flows along here like it did before.  And, thanks to TypeScript's flexible type system, the above is totally typesafe and will let you know the right types of things and will still error if you happen to make mistakes.
+
+So, how was this done?  Well, the core part of the change is thanks to a little-known feature of JavaScript: (Proxies)[https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy].
+
+Proxies allow you to return objects where you can control many facets of how JavaScript works at runtime.  There are many things a Proxy lets you control, however for our needs the most important bit was that it allows you to override how member-lookup works.  In other words, when you write `someProxy.someMember` you get a chance to determine what should happen an what `someMember` should actually return!  With that flexible interception point available, we actually took our core `Output` type and made it into a `Proxy`.  We then added the right interception code so that if you ever write `someOutput.someMember` that that gets translated exactly into `someOutput.apply(o => o.someMember)`.   his also works just fine for array-accesses (which are just property-lookups from JavaScript's perspective).  In other words, at runtime, the member-lookup form and the `.apply` form will be equivalent (except that the former is so much nicer to write!).  This is why, for example, `certCertificate.domainValidationOptions[0].resourceRecordValue` will have the correct value (with all the right dependency information).  At runtime it really will be equivalent to the original `certCertificate.apply(certCertificate => certCertificate.domainValidationOptions[0].resourceRecordValue)` form.
```

Comment:
What about this, for grammar (avoiding use of the second person):

"Proxies allow users to write JavaScript code that intercepts and overrides many facets of JavaScript runtime behavior."

Reply:
In general I think grammatically we should avoid the second person in this paragraph, since we're not actually instructing the reader to do something; we're describing that proxies allow *us* to overload aspects of JavaScript instead of *you* the user of `Proxy`, since in the end we are the ones handling the proxy.

Is that reasonable, you think?